data-prepper-api/src/main/java/org/opensearch/dataprepper/model/sink/AbstractSink.java

```diff
@@ -17,14 +17,25 @@
  * This class implements the Sink interface and records boilerplate metrics
  */
 public abstract class AbstractSink<T extends Record<?>> implements Sink<T> {
+    static private final String NUM_RETRIES = "num_retries";
+    static private final int NUM_DEFAULT_RETRIES = 600;
     protected final PluginMetrics pluginMetrics;
     private final Counter recordsInCounter;
     private final Timer timeElapsedTimer;
+    private final Thread retryThread;
 
     public AbstractSink(final PluginSetting pluginSetting) {
         this.pluginMetrics = PluginMetrics.fromPluginSetting(pluginSetting);
         recordsInCounter = pluginMetrics.counter(MetricNames.RECORDS_IN);
         timeElapsedTimer = pluginMetrics.timer(MetricNames.TIME_ELAPSED);
+        Integer numRetries = (Integer) pluginSetting.getAttributeFromSettings(NUM_RETRIES);
+        if (numRetries == null) {
+            numRetries = NUM_DEFAULT_RETRIES;
+        }
+        retryThread = new Thread(new SinkThread(this, numRetries));
```

Comment:
I think this class design could be re-worked some.

Data Prepper core should call `initialize()` on the sink.

The `AbstractSink` can then provide a `doInitialize()` method which sub-classes will implement. Thus, `AbstractSink` can implement `initialize()` to have the code you have here:

```
retryThread = new Thread(...);
if(!isReady()) {
  retryThread.start();
}
```

Reply:
I thought about that approach but derived Sink class's (like OpenSearch Sink) constructor would not have got called by the time Abstract Sink called initialized() on the OpenSearch Sink. So, calling initialize() in OpenSearchSInk before it's constructor got invoked doesn't make sense. Correct me if my understanding is wrong.